je-lesson-4-hibernate/src/main/java/dao/Dao.java

```diff
@@ -0,0 +1,89 @@
+package dao;
+
+import hibernateutil.HibernateUtil;
+import lombok.extern.slf4j.Slf4j;
+import org.hibernate.Session;
+import org.hibernate.SessionFactory;
+
+import javax.persistence.criteria.CriteriaBuilder;
+import javax.persistence.criteria.CriteriaQuery;
+import javax.persistence.criteria.Root;
+import java.util.List;
+
+@Slf4j
+public abstract class Dao<T> {
+    private final SessionFactory sessionFactory = HibernateUtil.getSessionFactory();
+
+    public void create(T entity) {
+        if (entity != null) {
+            log.debug("method : 'create( {} )", entity.getClass().getSimpleName());
+            try (Session session = sessionFactory.openSession()) {
+                session.beginTransaction();
+
+                session.save(entity);
+
+                session.getTransaction().commit();
+            }
+        }
+    }
+
+    public T read(long id, Class<T> tClass) {
+        T entity;
+        log.debug("method : 'read(long id, {} )", tClass.getSimpleName());
+        try (Session session = sessionFactory.openSession()) {
+            entity = session.get(tClass, id);
+        }
+        return entity;
+    }
+
+    public void update(T entity) {
+        if (entity != null) {
+            log.debug("method : 'update( {} )", entity.getClass().getSimpleName());
+            try (Session session = sessionFactory.openSession()) {
+                session.beginTransaction();
+
+                session.update(entity);
+
+                session.getTransaction().commit();
+            }
+        }
+    }
+
+    public boolean delete(long id, Class<T> tClass) {
+        T entity;
+        boolean result = false;
+        log.debug("method : 'delete(long id, {} )", tClass.getSimpleName());
+        try (Session session = sessionFactory.openSession()) {
+            entity = session.get(tClass, id);
+            if (entity != null) {
+                session.beginTransaction();
+
+                session.delete(entity);
+
+                session.getTransaction().commit();
+                result = true;
+            }
+        }
+        return result;
+    }
+
+    public List<T> readAll(Class<T> tClass) {
```

Comment:
it is better do not return list as null, try to return empty list if there are no results